django/middleware/csrf.py

```diff
@@ -227,15 +251,21 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                     return self._reject(request, REASON_INSECURE_REFERER)
 
                 # If there isn't a CSRF_COOKIE_DOMAIN, assume we need an exact
-                # match on host:port. If not, obey the cookie rules.
-                if settings.CSRF_COOKIE_DOMAIN is None:
-                    # request.get_host() includes the port.
-                    good_referer = request.get_host()
-                else:
+                # match on host:port. If not, obey the cookie rules (or those
+                # for the session cookie, if we use sessions).
+                if settings.CSRF_COOKIE_DOMAIN is not None and not settings.CSRF_USE_SESSIONS:
                     good_referer = settings.CSRF_COOKIE_DOMAIN
                     server_port = request.get_port()
                     if server_port not in ('443', '80'):
                         good_referer = '%s:%s' % (good_referer, server_port)
```

Comment:
Here we need
```python
  elif settings.CSRF_COOKIE_DOMAIN is not None and settings.CSRF_USE_SESSIONS:
      raise ImproperlyConfigured("When CSRF uses sessions, use SESSION_COOKIE_DOMAIN instead of CSRF_COOKIE_DOMAIN")
```
This, to make sure people don't just forget such settings, which probably express some security assumptions about their deployment.